README.md
# open-commons-json
Open Commons for JSON


history.md
[2022/04/07]
- Release: 1.8.0-SNAPSHOT
- Tag: 1.7.0
- Dependencies:
  + open.commons.core: 2.0.0-SNAPSHOT

[2022/04/07]
- Release: 1.7.0

[2021/12/03]
- Dependencies
  + open.commons.core: 1.8.0-SNAPSHOT

[2019/04/09]
- Release: 1.6.4
- Rebuild source directory structure.

[2019/03/29]
- Release: 1.6.3.4
- Update
 + open.commons.core updated: 1.6.5

[2018/09/10]
- Release: 1.6.3.3
- Update
 + open.commons.core updated: 1.4.17

[20148/04/30]
- Release: 1.6.3.2
- Update
 + open.commons.core updated: 1.4.16

[2017/12/13]
- Release: 1.6.3.1
- Update
 + open.commons.core updated: 1.4.10

[2017/11/29]
- Release: 1.6.3
- Add
 + open.commons.json.util.JSONUtil.toMapStartWith(String)
 + open.commons.json.util.JSONUtil.toMapStartWithout(String)

[2017/11/03]
- Release: 1.6.2.1
- Add
 + open.commons.json.util.JSONUtil.toMap(JSONObject): 내부 수정. HashMap -&gt; FIFOMap

[2017/10/20]
- Release: 1.6.2
- Add
 + open.commons.json.function.JSONFunction

[2017/09/22]
- Release: 1.6.1.5
- Update
 + open-commons-core updated: 1.4.7

[2017/09/12]
- Release: 1.6.1.4
- Update
 + open-commons-core updated: 1.4.6

[2017/09/11]
- Release: 1.6.1.3
- Update
 + open-commons-core updated: 1.4.5

[2017/08/21]
- Release: 1.6.1.2
- Update
 + open-commons-core updated: 1.4.2.1 -&gt; 1.4.3.1

[2017/07/06]
- Release: 1.6.1.1
- Update
 + open-commons-core updated: 1.4.2 -&gt; 1.4.2.1

[2016/11/01]
- Release: 1.6.1
- Update
 + open.commons.json.model.toOrderedJSONObject(JSONObject): 버그 수정

[2016/09/21]
- Release: 1.6.0.1
- Update
 + open-commons-core updated: 1.4.0 -&gt; 1.4.1 'properties ordering'

[2016/05/26]
- Release: 1.6.0
- Update
 + set 'properties ordering'

[2016/05/04]
- Release: 1.5.0
- Update
 + code of &quot;org.codehaus.jettison.json&quot;: 1.3.7

[2015/12/23]
- Release: 1.4.7
- Update
 + open-commons-core version: 1.4.0

[2015/12/23]
- Release: 1.4.5.3
- Update
 + open-commons-core version: 1.3.3

[2015/07/23]
- Update
 + open-commons-core version: 1.3.2

[2015/03/17]
- Release: 1.4.5.2
- Update
 + open-commons-core version: 1.3.0

[2015/03/12]
- Release: 1.4.5
- Update
 + open-commons-core version: 1.2.6

[2015/01/28]
- Release: 1.4.4

[2015/01/13]
- Release: 1.4.3.1
- Update
 + open-commons-core version: 1.2.2 &lt;- 1.2.1

[2015/01/06]
- Update
 + open-commons-core version: 1.2.1 &lt;- 1.2.0

[2014/12/29]
- Release
 + 1.4.3
- Add
 + JSONArrayFactory.toJSONArray(Collection&lt;T&gt;, Class&lt;T&gt;, GenericTypeVariable...)
 + JSONArrayFactory.toJSONArray(Collection&lt;T&gt;, Class&lt;T&gt;, List&lt;Class&lt;?&gt;&gt;)

 + JSONArrayFactory.toJSONArray(Collection&lt;T&gt;, Class&lt;T&gt;, Class&lt;?&gt;)

 + JSONArrayFactory.toJSONArray(Object, Class&lt;T&gt;, GenericTypeVariable...)

 + JSONArrayFactory.toJSONArray(Object, Class&lt;T&gt;, List&lt;Class&lt;?&gt;&gt;)

 + JSONArrayFactory.toJSONArray(Object, Class&lt;T&gt;, Class&lt;?&gt;)



[2014/11/25]
- Release
 + 1.4.1 &lt;- 1.3.1
- Update
 + Remove open.commons.json.model.AbstractJSONModel.hashCode()
- Bug Fix
 + Rebirth open.commons.json.model.AbstractJSONModel.hashCode()

[2014/11/11]
- Release
 + 1.3.1 &lt;- 1.3.0.1

[2014/11/07]
- Release
 + 1.3.0.1 &lt;- 1.3.0
 + open-commons-core version: 1.2.0 &lt;- 1.1.3

[2014/11/04]
- Release
 + 1.3.0 &lt;- 1.2.2.1
- Bug Fix
 + open.commons.json.model.JSONMapper.unmarshall(String, Object, BuildConfig, Object ...)에서 
    Object가 null 인 경우 처리

[2014/10/17]
- Release
 + 1.2.2.1 &lt;- 1.2.2
 + open-commons-core version: 1.1.3 &lt;- 1.1.2

[2014/09/26]
 + open-commons-core version: 1.1.2 &lt;- 1.1.1

[2014/09/17]
 + open-commons-core version: 1.1.1 &lt;- 1.1.0

[2014/09/10]
- Release: 1.2.2
- Change
 + open-commons-core version: 1.0.10 -&gt; 1.1.0

[2014/07/29]
- Release: 1.2.1
- Add
 + open.commons.jsonx.util

[2014/07/25]
- Add
 + public static &lt;T&gt; T open.commons.json.util.JSONUtil.get(JSONObject, String, T) throws JSONException;

[2014/07/10]
- Change
 + update open-commons-core version: 1.0.9 -&gt; 1.0.10

[2014/07/07]
- Release 1.0.10
- Patch
 + public static final &lt;T extends Serializable &amp; IJSONString&gt; JSONArray open.commons.json.model.JSONArrayFactory.toJSONArray(Collection&lt;T&gt; values, Class&lt;T&gt; elemType) throws JSONException {
        return toJSONArray(values.toArray(), elemType);
    } 
    의 구현 부분을 
     return toJSONArray(values.toArray(new IJSONString[]{}), elemType);로 수정
[2014/06/19]
- New
 + open.commons.json.model.exception.JSONMarshalledException
 + open.commons.json.model.exception.JSONUnmarshalledException

[2014/06/18]
- Release 1.0.9
 + Support variable json properties.

[2014/06/04]
- Release 1.0.8

- Changes !!!!!!!!!!!!!
 + open.commons.core-1.0.8에 따른 side effects 처리

[2014/06/03]
- New !!!
 + Support classes contains fields which are not a child of IJSONString.

[2014/05/30]
- Hot !!!
 + License is changed!!!
   - before: COPYLEFT by 'Open Commons' &amp;  Park Jun-Hong All Rights Reserved when use for commercial purpose.
   - after: 

[2014/05/08]
- Changed !!!
 +  open.commons.json.model.DefaultJSONModel.investigateField(Field, String) 를 open.commons.json.model.JSONMapper.investigateFiled(T, Field, String)로 이관.
- open.commons.json.model.JSONArrayFactory 에 JSONArray 생성 메소드 추가

[2014/05/07]
- open.commons.json.model.JSONArrayFactory
 + JSONArray 대신 String 를 파라미터로 받는 메소드들 추가.
 
[2014/05/02]
- open-commons-core version up: 1.0.5 -&gt; 1.0.6

# ....
 
1.5
- open.commons.json.model.JSONArrayFactory 추가.

1.4
 - Set a modifier of AbstractJSONModel.hashCode() as 'final'.

src/open/commons/json/IJSONString.java
/*

 * Copyright 2011 Park Jun-Hong (parkjunhong77/gmail/com)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*


 * 
 * 
 * This file is generated under this project, "open-commons-json".
 *
 * Date  : 2012. 11. 17. 오전 5:07:00
 *
 * Author: Park_Jun_Hong_(fafanmama_at_naver_com)
 *
 */
package open.commons.json;

import java.io.Serializable;
import java.lang.reflect.Field;
import java.util.List;

import open.commons.json.annotation.JSONField;
import open.commons.json.model.BuildConfig;
import open.commons.reflect.GenericTypeVariable;
import open.commons.utils.ConvertUtils;

import code.org.codehaus.jettison.json.JSONException;
import code.org.codehaus.jettison.json.JSONObject;

/**
 * JSON 문자열과 상호연동을 지원하는 인터페이스.
 * 
 * @author Park_Jun_Hong_(fafanmama_at_naver_com)
 * 
 */
public interface IJSONString extends Serializable, Cloneable {

    /**
     * {@link JSONField}가 추가된 모든 {@link Field}의 값을 초기화 시킨다.
     * 
     * @see ConvertUtils#resetField(Object, Field)
     */
    public void clear();

    // public BuildConfig getBuildConfig();

    /**
     * Not yet supported.
     * 
     * @param jsonModel
     * 
     * @since 2014. 5. 30.
     */
    public void inject(IJSONString jsonModel);

    public IJSONString mature(JSONObject json) throws JSONException;

    public IJSONString mature(JSONObject json, BuildConfig buildConfig) throws JSONException;

    public IJSONString mature(JSONObject json, Class<?> paramType) throws JSONException;

    public IJSONString mature(JSONObject json, GenericTypeVariable... paramTypes) throws JSONException;

    public IJSONString mature(JSONObject json, List<Class<?>> paramTypes) throws JSONException;

    public IJSONString mature(String jsonString) throws JSONException;

    public IJSONString mature(String jsonString, BuildConfig buildConfig) throws JSONException;

    public IJSONString mature(String jsonString, Class<?> paramType) throws JSONException;

    public IJSONString mature(String jsonString, GenericTypeVariable... paramTypes) throws JSONException;

    public IJSONString mature(String jsonString, List<Class<?>> paramTypes) throws JSONException;

    public JSONObject toJSONObject() throws JSONException;

    /**
     * 
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26
     * @since 1.6.0
     */
    public JSONObject toJSONObject(boolean ordered) throws JSONException;

    /**
     * 
     * @param ordered
     * @param paramTypes
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26
     * @since 1.6.0
     */
    public JSONObject toJSONObject(boolean ordered, GenericTypeVariable... paramTypes) throws JSONException;

    public JSONObject toJSONObject(BuildConfig buildConfig) throws JSONException;

    /**
     * 
     * @param buildConfig
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26
     * @since 1.6.0
     */
    public JSONObject toJSONObject(BuildConfig buildConfig, boolean ordered) throws JSONException;

    public JSONObject toJSONObject(Class<?> paramType) throws JSONException;

    /**
     * 
     * @param paramType
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26
     * @since 1.6.0
     */
    public JSONObject toJSONObject(Class<?> paramType, boolean ordered) throws JSONException;

    public JSONObject toJSONObject(GenericTypeVariable... paramTypes) throws JSONException;

    public JSONObject toJSONObject(List<Class<?>> paramTypes) throws JSONException;

    /**
     * 
     * @param paramTypes
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26
     * @since 1.6.0
     */
    public JSONObject toJSONObject(List<Class<?>> paramTypes, boolean ordered) throws JSONException;

    public String toJSONString() throws JSONException;

    /**
     * 
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(boolean ordered) throws JSONException;

    /**
     * 
     * @param ordered
     * @param paramTypes
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     */
    public String toJSONString(boolean ordered, GenericTypeVariable... paramTypes) throws JSONException;

    public String toJSONString(BuildConfig buildConfig) throws JSONException;

    /**
     * 
     * @param buildConfig
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(BuildConfig buildConfig, boolean ordered) throws JSONException;

    public String toJSONString(BuildConfig buildConfig, int factor) throws JSONException;

    /**
     * 
     * @param buildConfig
     * @param factor
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(BuildConfig buildConfig, int factor, boolean ordered) throws JSONException;

    public String toJSONString(Class<?> paramType) throws JSONException;

    /**
     * 
     * @param paramType
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(Class<?> paramType, boolean ordered) throws JSONException;

    public String toJSONString(GenericTypeVariable... paramTypes) throws JSONException;

    public String toJSONString(int factor) throws JSONException;

    /**
     * 
     * @param factor
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(int factor, boolean ordered) throws JSONException;

    /**
     * 
     * @param factor
     * @param ordered
     * @param paramTypes
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(int factor, boolean ordered, GenericTypeVariable... paramTypes) throws JSONException;

    public String toJSONString(int factor, Class<?> paramType) throws JSONException;

    /**
     * 
     * @param factor
     * @param paramType
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(int factor, Class<?> paramType, boolean ordered) throws JSONException;

    public String toJSONString(int factor, GenericTypeVariable... paramTypes) throws JSONException;

    public String toJSONString(int factor, List<Class<?>> paramTypes) throws JSONException;

    /**
     * 
     * @param factor
     * @param paramTypes
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(int factor, List<Class<?>> paramTypes, boolean ordered) throws JSONException;

    public String toJSONString(List<Class<?>> paramTypes) throws JSONException;

    /**
     * 
     * @param paramTypes
     * @param ordered
     * @return
     * @throws JSONException
     *
     * @since 2016. 5. 26.
     * @since 1.6.0
     */
    public String toJSONString(List<Class<?>> paramTypes, boolean ordered) throws JSONException;
}
